[runSofa] improve the html DocBrowser

damienmarchal opened the pull request:
Small improvement on the view of html file associated with scenes. We can now navigate (with backward) between and starting Sofa scene directly with hyperlinks. The windows can also be hidden/displayed and its location preserved. 

I cannot prevent myself to also clean small things in RealGUI and SetDirectory/FileSystem. 
But this is very minor. 

I made a quick & dirty video to show the results so that you don't need to compile it: 
https://youtu.be/5uumC9f-wzo


______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

damienmarchal committed:
[SofaKernel] Adds a getExtension function to FileSystem.

There was one in SetDirectory but I see ne reason to have that here.
So for consistency I move it in FileSystem.

Backward compatibility is preserved.

damienmarchal committed:
[SofaKernel] Adds a function to get the list of file extension supported by the "static" SceneLoaderFactory.

damienmarchal committed:
[runSofa] add a ressources directory to hold runSofa specific data.

damienmarchal committed:
[runSofa] Adds a GuiDataRepository to manage runSofa specific ressources.

damienmarchal committed:
[runSofa] Update GraphiListenerQListView to take into account the file moved.

damienmarchal committed:
[runSofa] Add a DocBrowser window to view in better condition the docs.

Refactor the way the html documentation associated with Sofa scene
are handled. The doc browser now have an history and link to sofa
scene are correctly handled to load the sofa scene. This is very
practical to make self documenting libraries.

The DocBrowser can be displayed from the Help menu.

damienmarchal committed:
[runSofa]: fix typos.

damienmarchal committed:
[SofaKernel] Move some functon from .h to .cpp to avoid extra inclusions.

guparan commented:
It would be nice to have this in v17.12
If [ci-build] is OK, let's merge :-)

guparan commented on applications/sofa/gui/qt/GuiDataRepository.cpp:
```diff
@@ -0,0 +1,38 @@
+{
+namespace qt
+{
+    sofa::helper::system::FileRepository GuiDataRepository("RUNSOFA_DATA_PATH", "../share/sofa/runSofa/ressources/");
```
Why not using current resources dirs (e.g. share/icons)?
BTW, only one S for **resources** ;-)

damienmarchal replied:
Good question.

This is to make things modular as I see no reason why the icons specific to a given end user application (here runSofa) should be in the data directory of the base framework. Said differently, If the user don't care about runSofa (or runSofa2) why its data repository should be polluated by these resources ? 

So each project/plugin specific resources should reside in its own directory not in share/

guparan replied:
I agree with the modular idea but SofaGuiQt is not runSofa so I think it would better be `sofa::helper::system::FileRepository runSofaDataRepository("RUNSOFA_DATA_PATH", "../share/sofa/runSofa/ressources/");` created in applications/projects/runSofa or `sofa::helper::system::FileRepository GuiDataRepository("GUI_DATA_PATH", "../share/sofa/gui/ressources/");` in applications/sofa/gui.
What do you think?

damienmarchal replied:
Plugins examples...are in the example sub-directory of the plugin directory and not in share. 
Icon sets of external UI 'project' (like runSofa2) are in their project directory. 
Actually even before this PR the icons of runSofa were not in share but in the runSofa dir...so please don't change that and let's keep a single rule that is clear to understand: specific per 'project' (or 'plugins') files have to stay in their project (plugin) directory.

guparan replied:
Yep, that's exactly what I was saying in my second post. So what about my proposal?

damienmarchal replied:
Sorry I miss-understood you last comment. 

I see no problem in making things more consistent by renaming things as long as the files stays separated from the sofa/share.
Can you do that..caus .I'm focused on something else.

guparan commented:
[ci-build] (testing)

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal commented:
[ci-build]

damienmarchal commented:
[ci-build][with-scene-tests] do you really work ?

guparan commented:
Yep, I just fixed it ;)

damienmarchal committed:
[runSofa] Add a new directory to hold ressources for runSofa.

damienmarchal committed:
[runSofa] Fix the different ressources files & path.

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal committed:
[runSofaHQ] rename ressource into resource

Because Sofa is such an HQ software that we cannot live with mispelling in
our pathnames ;)

guparan commented:
[ci-build][with-scene-tests]

damienmarchal commented:
[ci-build][with-scene-tests]

guparan commented:
@damienmarchal Did you notice the unit test failures? It's about PSL.

damienmarchal commented:
@guparan thank you for pointing that.The problem is that the CI machine *_pr3 does not have the python module 'hjson' installed consequently psl tests failed.

EDIT: So you can safely ignore them...& merge the PR.

guparan commented:
Ok I'm gonna fix that, I'll also fix the missing timeout command on the mac_pr3 VM.

guparan commented:
Done! [ci-build][with-scene-tests]

damienmarchal commented:
The fails again... the mac_clang-3.4 succeeded (I restarted but the previous build was here
https://ci.inria.fr/sofa-ci/job/mac_clang-3.4_options_pr3/25/console)
And the additional failure in test scenes is because some machines still don't have the libjpeg/tiff library installed (se macos pr3 ) :) 

@guparan, @hugtalbot  if you are passing by please considere merging the PR because we need it monday morning for our group writing of tutorials for Sofa :)
